### PR TITLE
Rails 6.1 fixes

### PR DESCRIPTION
**Pull request by dr-itz**
Note: one PG test still fails

### CODE REVIEW

**Comment by enebo** [comment]
@dr-itz There is only a single failure involving pluck and some index being the value 1 vs "1".   Is that followup work as I did not see if in the last CI run as failing?

**Comment by enebo** [comment]
@rdubya @kares eef5f826391413dda238ea23c9526b580c82776d looks tasty to me but I would like a couple more eyes on that change.

**Comment by dr-itz** [comment]
I looked into the pluck test. In `PostgreSQLResult.columnTypeMap` we create a map of column name => type. This works fine as long as column names are unique. But that failing has

```
Arel.sql("COALESCE(id, 0)"),
Arel.sql("COALESCE(title, 'untitled')")
```

i.e. w/o a column alias and for both column labels we get "coalesce"...second one wins.

**Comment by dr-itz** [comment]
> I looked into the pluck test. In `PostgreSQLResult.columnTypeMap` we create a map of column name => type. 

looking at https://github.com/rails/rails/blob/master/activerecord/lib/active_record/result.rb#L106
it also supports an array. But that crashes when trying to load the schema. Looking into that now

**Comment by dr-itz** [comment]
> looking at https://github.com/rails/rails/blob/master/activerecord/lib/active_record/result.rb#L106
> it also supports an array. But that crashes when trying to load the schema. Looking into that now

ignore me, that's for overrides only

File: src/java/arjdbc/jdbc/RubyJdbcConnection.java [hunk 1]
```diff
@@ -3627,7 +3656,10 @@ protected final TableName extractTableName(
     }
 
     protected IRubyObject valueForDatabase(final ThreadContext context, final IRubyObject attribute) {
-        return attribute.callMethod(context, "value_for_database");
+        if (attribute.respondsTo("value_for_database")) {
```
Comment (kares) [inline src/java/arjdbc/jdbc/RubyJdbcConnection.java:3659]: would like to understand, why is this necessary - what would not respond_to? `value_for_database`?
asking since it's now double the method lookup cost and I wonder if this is a hot spot.

**Comment by dr-itz** [comment]
> I haven't been keeping up with the latest ActiveRecord changes, do they support passing in primitives? Last I knew everything had to be an attribute.

Unfortunately, that's a thing now:

```ruby
        def type_casted_binds(binds)
          case binds.first
          when Array
            binds.map { |column, value| type_cast(value, column) }
          else
            binds.map do |value|
              if ActiveModel::Attribute === value
                type_cast(value.value_for_database)
              else
                type_cast(value)
              end
            end
          end
        end
```

the array case is the legacy thing we on the ruby side. 

we might well do the same check for `ActiveModel::Attribute` and then do all the thing we did until now and only use the new thing otherwise

**Comment by dr-itz** [comment]
> we might well do the same check for `ActiveModel::Attribute` and then do all the thing we did until now and only use the new thing otherwise

just did that on top. Also fixed the datetime thing and a few style issues. And excluded the failing PG test for now

**Comment by enebo** [comment]
Wouldn't raw values be faster than wrapping it in an attribute object?  Am I missing something?  I took this as a big possible improvement in removing an extra object boxing and indirection which would need to be inlined.

**Comment by dr-itz** [comment]
Not wrapping is nice...and just for fun, I changed one line in `setStatementParameter` to always use the old `ActiveModel::Attribute` code paths. That's how a rails test run for SQLite3 looks like:

    6802 runs, 16328 assertions, 15 failures, 598 errors, 14 skips

so not wrapping seems to become a thing

File: src/java/arjdbc/jdbc/RubyJdbcConnection.java (hunk 1 above)
Comment (dr-itz) [reply src/java/arjdbc/jdbc/RubyJdbcConnection.java]: gone with the latest version

**Comment by kares** [comment]
great, another full circle from not wrapping to wrapping and now back to not wrapping ... still the "same old" AR :wilted_flower:

**Comment by enebo** [comment]
@dr-itz sorry I should have merged this a week ago.  Thanks again for keeping our head above water.
